ecs-agent/credentials/providers/instance_credentials_provider_windows.go

```diff
@@ -0,0 +1,104 @@
+//go:build windows
+// +build windows
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package providers
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials/ec2rolecreds"
+)
+
+// The instance credentials chain is the default credentials chain plus the "rotating shared credentials provider",
+// so credentials will be checked in this order:
+//
+//  1. Env vars (AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY).
+//
+//  2. Shared credentials file (https://docs.aws.amazon.com/ses/latest/DeveloperGuide/create-shared-credentials-file.html) (file at ~/.aws/credentials containing access key id and secret access key).
+//
+//  3. EC2 role credentials. This is an IAM role that the user specifies when they launch their EC2 container instance (ie ecsInstanceRole (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/instance_IAM_role.html)).
+//
+//  4. Rotating shared credentials file located at /rotatingcreds/credentials
+//
+//     The default credential chain provided by the SDK includes:
+//     * EnvProvider
+//     * SharedCredentialsProvider
+//     * RemoteCredProvider (EC2RoleProvider)
+//
+//     In the case of ECS-A on Windows, the `SharedCredentialsProvider` takes
+//     precedence over the `RotatingSharedCredentialsProvider` and this results
+//     in the credentials not being refreshed. To mitigate this issue, we will
+//     reorder the credential chain and ensure that `RotatingSharedCredentialsProvider`
+//     takes precedence over the `SharedCredentialsProvider` for ECS-A.
+func NewInstanceCredentialsProvider(
+	isExternal bool,
+	rotatingSharedCreds aws.CredentialsProvider,
+	imdsClient ec2rolecreds.GetMetadataAPIClient,
+) *InstanceCredentialsProvider {
+	var providers []aws.CredentialsProvider
+
+	// If imdsClient is nil, the SDK will default to the EC2 IMDS client.
+	// Pass a non-nil imdsClient to stub it out in tests.
+	options := func(o *ec2rolecreds.Options) {
+		o.Client = imdsClient
+	}
+
+	if isExternal {
+		providers = []aws.CredentialsProvider{
```

Comment:
Do all of these explicitly configured providers need to be wrapped with type [CredentialsCache](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#CredentialsCache) to ensure each of them individually cache and are concurrency safe? Or does only having the top level provider specified in `agent/app/agent.go` (which contains all these individual providers) wrapped suffice?

I ask because I saw this in the [docs](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specify-credentials-programmatically):
> All credential providers passed to or returned by `LoadDefaultConfig` are wrapped in a [CredentialsCache](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#CredentialsCache) automatically. This enables caching, and credential rotation that is concurrency safe. If you explicitly configure a provider on `aws.Config` directly you must also explicitly wrap the provider with this type using [NewCredentialsCache](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#NewCredentialsCache).

`defaultCreds(options)` on L70 seems to be covered since it uses a credential provider returned by `LoadDefaultConfig`, but how about all the other providers?

Reply:
Great question! I made this test program to determine the behavior. https://gist.github.com/tinnywang/3ce5483633a723e1e11050665362c13e

`CredsProvider` wraps `envCreds`, which retrieves creds from env vars. Any time `envCreds` retrieves creds, it prints "retrieving creds from env config".


Commenting out L17 and uncommenting L18 wraps `CredsProvider` in a cache. When creds are repeatedly fetched from the cache, the output is
```
2024/11/14 16:43:13 retrieving creds from env config
2024/11/14 16:43:13 ****************EW64
2024/11/14 16:43:15 ****************EW64
2024/11/14 16:43:17 ****************EW64
2024/11/14 16:43:19 ****************EW64
...
```
The line "retrieving creds from env config" is only logged once, which means `envCreds.Retrieve` is only called once even though `provider.Retrieve` is called every iteration.

Uncommenting L17 and commenting out L18 makes the program use the standalone `CredsProvider`. In this case, the output is
```
2024/11/14 16:53:07 retrieving creds from env config
2024/11/14 16:53:07 ****************EW64
2024/11/14 16:53:09 retrieving creds from env config
2024/11/14 16:53:09 ****************EW64
2024/11/14 16:53:11 retrieving creds from env config
2024/11/14 16:53:11 ****************EW64
...
```
This means `envCred.Retrieve` is called every time `provider.Retrieve` is called.

So it looks like wrapping the toplevel credentials provider in a cache will prevent the underlying providers from unnecessarily fetching creds.

Reply:
Great, thanks much for confirming this and the helpful programmatic demonstration!